lib/rules/avoid-leaking-state-in-ember-objects.js

```diff
@@ -0,0 +1,66 @@
+'use strict';
+
+const ember = require('../utils/ember');
+
+//------------------------------------------------------------------------------
+// Ember object rule - Avoid leaking state
+// (Don't use arrays or objects as default props)
+//------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Avoids state leakage',
+      category: 'Ember Object',
+      recommended: true
+    },
+    fixable: null, // or "code" or "whitespace"
+    schema: [{
+      type: 'array',
+      items: { type: 'string' },
+    }],
+  },
+
+  create(context) {
+    let ignoredProperties = context.options[0] || [];
+
+    const message = {
+      array: 'Avoid using arrays as default properties',
+      object: 'Avoid using objects as default properties',
+    };
+
+    const report = function (node, messageType) {
+      context.report(node, message[messageType]);
+    };
+
+    ignoredProperties = ignoredProperties.concat([
+      'classNames',
+      'classNameBindings',
+      'actions',
+      'concatenatedProperties',
+      'mergedProperties',
+      'positionalParams',
+      'attributeBindings',
+      'queryParams',
+      'attrs',
+    ]);
+
+    return {
+      CallExpression(node) {
+        if (!ember.isEmberObject(node)) return;
+
+        const properties = ember.getModuleProperties(node);
+
+        properties
+          .filter(property => ignoredProperties.indexOf(property.key.name) === -1)
+          .forEach((property) => {
+            if (ember.isObjectProp(property)) {
```

Comment:
I think it would better for us to list known immutable objects here instead of blacklisting objects and arrays.  Basically, I think that we should whitelist strings, numbers, symbols, boolean, null, undefined, and function. Everything else would be an error...

Reply:
Consider:

```js
Foo.extend({
  derp: new Set()
})
```

Reply:
that's a good idea but I think we should not whitelist undefined and null as sometimes object could be initialize as null or undefined and later could hold any other kind of object.
what do you think?

Reply:
We discussed this out of band, and I believe straightened out any misunderstandings.  I still believe that we should use a whitelist approach (allowing immutable primitives and function only).